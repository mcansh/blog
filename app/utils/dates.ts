import { formatISO, parseISO, format } from 'date-fns';

type PossibleDates = string | number | Date;

function parseDate(input: PossibleDates) {
  return typeof input === 'string' ? parseISO(input) : input;
}

const formatPostDate = (inputDate: PossibleDates) =>
  format(parseDate(inputDate), 'MMMM d, yyyy');

const iso8601 = (date: PossibleDates) =>
  formatISO(parseDate(date), { representation: 'date' });

export { iso8601, formatPostDate };
