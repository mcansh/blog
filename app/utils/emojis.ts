const emojis = [
  '😁',
  '🤙',
  '🙌',
  '🍻',
  '🔥',
  '😬',
  '👍',
  '🐬',
  '👀',
  '👌',
  '😉',
  '💪',
  '💩',
  '🎉',
  '💣',
  '😘',
  '😜',
  '🙏',
  '👋',
  '💎',
  '🚀',
  '💰',
];

function randomEmoji() {
  return emojis[Math.floor(Math.random() * emojis.length)];
}

export { emojis };
export default randomEmoji;
