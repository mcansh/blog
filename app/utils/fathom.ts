import * as React from 'react';
import * as Fathom from 'fathom-client';
import { useLocation } from 'react-router';

function useFathom(siteId: string, url?: string) {
  const { pathname } = useLocation();

  React.useEffect(() => {
    Fathom.load(siteId, {
      url: url ? url : undefined,
      excludedDomains: ['localhost'],
    });
  }, [siteId, url]);

  React.useEffect(() => {
    Fathom.trackPageview();
  }, [pathname]);
}

export { useFathom };
