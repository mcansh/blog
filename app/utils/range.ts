function range(start: number, end: number) {
  if (start > end) {
    throw new RangeError('end cannot be less than start');
  }

  return Array(end - start + 1)
    .fill(null)
    .map((_, idx) => start + idx);
}

export { range };
