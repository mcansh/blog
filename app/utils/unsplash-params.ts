const unsplashParams = (url?: string | undefined): string => {
  if (url == null) return '';
  return url.startsWith('https://unsplash.com')
    ? `${url}?utm_source=unsplash&utm_medium=referral&utm_content=mcansh_blog`
    : url;
};

export default unsplashParams;
