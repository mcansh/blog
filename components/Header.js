import React from 'react';
import PropTypes from 'prop-types';
import { H2 } from './post/Typography';
import Button from './Button';

class Header extends React.Component {
  constructor(props) {
    super(props);
    this.state = { blur: 'blur(0px)' };
    this.handleScroll = this.handleScroll.bind(this);
  }

  componentDidMount() {
    window.addEventListener('scroll', this.handleScroll, true);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll, true);
  }

  handleScroll() {
    const scroll = window.pageYOffset;
    const maxMinBlur = Math.min(100, Math.max(0, scroll / 3));
    const blur = `blur(${maxMinBlur}px)`;
    this.setState({ blur });
  }
  render() {
    const Date = this.props.date ? <H2>{this.props.date}</H2> : '';
    const blurStyles = {
      backdropFilter: `${this.state.blur}`,
      WebkitBackdropFilter: `${this.state.blur}`,
    };
    return (
      <header>
        <div className="header__bg" style={{ backgroundImage: `linear-gradient(0deg, rgba(0, 0, 0, 0.74902), rgba(0, 0, 0, 0.74902)), url(/static/images/${this.props.image})` }} />
        <div className="header__content" style={blurStyles}>
          <h1>{this.props.text}</h1>
          {/* {buttonLink(this.props.slug)} */}
          <Button link={this.props.slug}>Read More</Button>
          {Date}
        </div>
        <style jsx>{`
          header {
            height: 50vh;
            min-height: 500px;
            display: flex;
            justify-content: center;
            align-items: center;
            position: relative;
            font-size: 1.7em;
            text-align: center;
          }

          .header__bg {
            position: absolute;
            top: 0;
            right: 0;
            bottom: 0;
            left: 0;
            background-position: center;
            background-size: cover;
          }

          .header__content {
            color: white;
            z-index: 1;
            height: 100%;
            width: 100%;
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;
          }

          @supports (backdrop-filter: initial) and (-webkit-backdrop-filter: initial) {
            .header_content {
              will-change: backdrop-filter;
              will-change: webkit-backdrop-filter;
            }
          }

          h1, h2 {
            max-width: 80%;
          }

          @media (max-width: 410px) {
            h1,h2 {
              max-width: 90%;
            }
            h1 {
              font-size: 1.5em;
            }
          }

          @media (max-height: 500px) {
            header {
              height: 100vh;
              min-height: 100vh;
            }
          }

          @media (max-width: 500px) {
            header {
              min-height: 95vh;
              height: 95vh;
            }
          }
        `}</style>
      </header>
    );
  }
}

Header.defaultProps = {
  date: '',
  slug: '',
};

Header.propTypes = {
  text: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  date: PropTypes.string,
  slug: PropTypes.string,
};


export default Header;
