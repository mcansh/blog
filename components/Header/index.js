// @flow
import React from 'react';
import styled from 'styled-components';
import { Spring, animated } from 'react-spring';
import FindPost from '../../utils/findPost';
import Button from '../Button';
import Image from './Image';
import Date from './Date';

const HeaderStyles = styled.header.attrs({ 'data-testid': 'header' })`
  height: 50vh;
  min-height: 50rem;
  max-height: 80rem;
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;
  position: relative;

  @media (orientation: landscape) and (max-height: 500px) {
    max-height: 100vh;
    min-height: 100vh;
    height: 100vh;
  }
`;

const HeaderContent = styled.div`
  z-index: 1;
  text-align: center;
  max-width: 80vw;
`;

export const Title = styled.h1`
  margin-bottom: 2rem;
  font-size: 3rem;

  @media (min-width: 400px) {
    font-size: 4rem;
  }
`;

const AnimatedTitle = animated(Title);

export type ImageTypes = {
  imageUrl: string,
  name?: string | null,
  url?: string | null,
};

type Props = {
  title?: string | null,
  link?: string | null,
  id?: string | null,
  image?: ImageTypes | null,
};

const Header = ({ image, title, link, id }: Props) => (
  <FindPost id={id}>
    {post => {
      const headerTitle = title != null ? title : post.title;
      const headerImage = image != null ? image : post.image;
      const showDate = link == null && post?.date;
      const showLink = link != null && post?.id;
      return (
        <HeaderStyles>
          <HeaderContent>
            <Spring
              native
              from={{ opacity: 0, transform: 'translateY(-50px)' }}
              to={{ opacity: 1, transform: 'translateY(0px)' }}
              config={{ duration: 400 }}
            >
              {props => (
                <AnimatedTitle style={props}>{headerTitle}</AnimatedTitle>
              )}
            </Spring>
            {showDate && <Date date={post.date} />}
            {showLink && <Button text="Read More" link={post.id} />}
          </HeaderContent>
          <Image image={headerImage} />
        </HeaderStyles>
      );
    }}
  </FindPost>
);

Header.defaultProps = {
  link: null,
  id: null,
  title: null,
  image: null,
};

export default Header;
