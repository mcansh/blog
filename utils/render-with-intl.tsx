/* eslint-env jest */
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from 'react-testing-library';
import theme from '~/config';

const renderWithIntl = (children: React.ReactNode) =>
  render(
    <ThemeProvider theme={theme}>
      <>{children}</>
    </ThemeProvider>
  );

export * from 'react-testing-library';
export { renderWithIntl as render };

export default renderWithIntl;
