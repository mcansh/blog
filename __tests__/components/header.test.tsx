/* eslint-env jest */
import React from 'react';
import 'jest-styled-components';
import { render } from '../../utils/render-with-intl';
import Header from '../../components/header';

describe('Header Component', () => {
  it('shows custom text and image', () => {
    const { getByTestId } = render(
      <Header
        title="BLOG"
        image={{
          imageUrl: '1*Wmv8hfi_bTHuHyV5CawnCw.jpg',
        }}
      />
    );

    const header = getByTestId('header');

    expect(header.textContent).toEqual('BLOG');

    expect(header.querySelector('img')).toHaveAttribute(
      'src',
      'https://res.cloudinary.com/dof0zryca/image/upload/f_auto/v1541889199/blog/1*Wmv8hfi_bTHuHyV5CawnCw.jpg'
    );
  });
});
