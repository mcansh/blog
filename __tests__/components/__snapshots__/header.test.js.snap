// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component with Snapshot Testing shows custom text and image 1`] = `
.c3 {
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
}

.c3::before {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(0,0,0,0.6);
}

.c4 {
  height: 100%;
  width: 100%;
  object-fit: cover;
}

.c0 {
  height: 50vh;
  min-height: 50rem;
  max-height: 80rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: white;
  position: relative;
}

.c1 {
  z-index: 1;
  text-align: center;
  max-width: 80vw;
}

.c2 {
  font-size: 5rem;
  margin-bottom: 2rem;
}

@media (orientation:landscape) and (max-height:500px) {
  .c0 {
    max-height: 100vh;
    min-height: 100vh;
    height: 100vh;
  }
}

<header
  className="c0"
>
  <div
    className="c1"
  >
    <h1
      className="c2"
    >
      BLOG
    </h1>
  </div>
  <picture
    className="c3"
  >
    <source
      srcSet="/static/images/posts/http://localhost:3000/static/images/posts/1*Wmv8hfi_bTHuHyV5CawnCw.webp"
      type="image/webp"
    />
    <source
      srcSet="/static/images/posts/http://localhost:3000/static/images/posts/1*Wmv8hfi_bTHuHyV5CawnCw.jpg"
      type="image/jpeg"
    />
    <img
      alt=""
      className="c4"
      data-source-url=""
      src="/static/images/posts/http://localhost:3000/static/images/posts/1*Wmv8hfi_bTHuHyV5CawnCw.jpg"
    />
  </picture>
</header>
`;

exports[`Header component with Snapshot Testing shows latest post 1`] = `
.c4 {
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
}

.c4::before {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(0,0,0,0.6);
}

.c5 {
  height: 100%;
  width: 100%;
  object-fit: cover;
}

.c3 {
  font-size: 3rem;
}

.c0 {
  height: 50vh;
  min-height: 50rem;
  max-height: 80rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: white;
  position: relative;
}

.c1 {
  z-index: 1;
  text-align: center;
  max-width: 80vw;
}

.c2 {
  font-size: 5rem;
  margin-bottom: 2rem;
}

@media (orientation:landscape) and (max-height:500px) {
  .c0 {
    max-height: 100vh;
    min-height: 100vh;
    height: 100vh;
  }
}

<header
  className="c0"
>
  <div
    className="c1"
  >
    <h1
      className="c2"
    >
      Ruby project
    </h1>
    <h2
      className="c3"
      title="December 9, 2017"
    >
      Posted 
      <span>
        8 months ago
      </span>
    </h2>
  </div>
  <picture
    className="c4"
  >
    <source
      srcSet="/static/images/posts/stacy-wyss-702.webp"
      type="image/webp"
    />
    <source
      srcSet="/static/images/posts/stacy-wyss-702.jpg"
      type="image/jpeg"
    />
    <img
      alt="Taken by Stacy Wyss"
      className="c5"
      data-source-url="https://unsplash.com/photos/-aDl1z8_nGY?utm_source=unsplash&utm_medium=referral&utm_content=mcansh_blog"
      src="/static/images/posts/stacy-wyss-702.jpg"
    />
  </picture>
</header>
`;
